torch/csrc/distributed/c10d/TraceUtils.h

```diff
@@ -495,16 +492,58 @@ struct NCCLTraceBuffer {
     return result;
   }
 
-  void retire_id(c10::optional<size_t> id) {
+  // Don't set compute_duration=True from inside dump API,
+  // becuase getDurationFromFirstEvent may hang and block dumping.
+  // OK to call from inside watchdog thread, which is probably going to hang
+  // anyway in cases where eventQuery will hang. (Then we fall back to rely on
+  // monitor thread to kill and dump, and rely on dump not hanging)
+  // But, don't hang holding the lock that dump() also holds!
+  void retire_id(c10::optional<size_t> id, bool compute_duration = false) {
     if (!enabled_ || !id) {
       return;
     }
-    std::lock_guard<std::mutex> guard(mutex_);
-    auto& entry = entries_.at(*id % max_entries_);
-    if (entry.id_ == *id) {
-      update_state(entry);
-      entry.retired_ = true;
-      entry.start_ = entry.end_ = nullptr;
+
+    bool can_compute_duration = false;
+    EventList* startEvents = nullptr;
+    EventList* endEvents = nullptr;
+
+    {
+      std::lock_guard<std::mutex> guard(mutex_);
+      auto& entry = entries_.at(*id % max_entries_);
+      if (entry.id_ == *id) {
+        update_state(entry);
+
+        if (compute_duration) {
+          can_compute_duration = strcmp(entry.state_, "completed") == 0 &&
+              entry.start_ && entry.end_;
+          startEvents = entry.start_;
+          endEvents = entry.end_;
+        }
+      }
+    }
+
+    // Compute duration without without holding the lock, because
+    // cudaEventDuration() can hang, and we need to acquire the lock before we
+    // can dump(), which we never want to block.
+    c10::optional<float> duration = c10::nullopt;
+    if (can_compute_duration) {
+      duration = getDurationFromFirstEvent(*startEvents, *endEvents);
```

Comment:
the other way to do this that i thought about, was to make it use the async-launcher pattern.  Unlike the other uses, this would happen frequently so i wouldn't want to make a new thread per duration request, we'd need some long lived worker thread.  I might do this later anyway, and have it also service other types of requests like we were discussing.  But maybe for now keep it simple?